Core/clim-basic/ports.lisp

```diff
@@ -171,107 +176,119 @@
-               (dispatch-event s new-event))))
-         (distribute-exit-events (sheet-b sheet-t event)
-           (when (and sheet-t sheet-b)
-             (do ((s sheet-b (sheet-parent s)))
-                 ((or (null s) (graftp s) (eq s sheet-t)))
-               (let ((new-event (shallow-copy-object event)))
-                 (get-pointer-position (s event)
-                   (setf (slot-value new-event 'x) x
-                         (slot-value new-event 'y) y
-                         (slot-value new-event 'sheet) s))
-                 (change-class new-event 'pointer-exit-event)
-                 (dispatch-event s new-event))))))
-    (let* ((pointer-sheet (get-pointer-event-sheet (event-sheet event) event))
-           (old-pointer-sheet (or (port-pointer-sheet port) (event-sheet event)))
-           (common-sheet (sheet-common-ancestor old-pointer-sheet pointer-sheet)))
-      (distribute-exit-events old-pointer-sheet common-sheet event)
-      (distribute-enter-events pointer-sheet common-sheet event)
-      (setf (port-pointer-sheet port) pointer-sheet)
-      ;; set the pointer cursor
-      (when pointer-sheet
-        (let ((pointer-cursor (sheet-pointer-cursor pointer-sheet)))
-	  (unless (eql (port-lookup-current-pointer-cursor port (event-sheet event))
-		       pointer-cursor)
-	    (set-sheet-pointer-cursor port (event-sheet event) pointer-cursor))))))
-  ;; Do not send the event twice.
-  (unless (typep event '(or pointer-enter-event pointer-exit-event))
-    (call-next-method)))
-
-;;; This method is necessary to distribute events for sheets which do
-;;; not have a mirror. In this case we find the innermost sheet at
-;;; event's position and dispatch event's shallow copy to this
-;;; sheet. -- jd 2019-08-05
+                    (t (setf sheet-a (sheet-parent sheet-a)))))))
+    (let ((common-ancestor (sheet-common-ancestor old-pointer-sheet
+                                                  new-pointer-sheet)))
+      ;; distribute exit events (innermost first)
+      (when (and common-ancestor old-pointer-sheet)
+        (do ((s old-pointer-sheet (sheet-parent s)))
+            ((or (null s) (graftp s) (eq s common-ancestor)))
+          (dispatch-event-copy s event 'pointer-exit-event)))
+      ;; distribute enter events (innermost last)
+      (dolist (s (do ((s new-pointer-sheet (sheet-parent s))
+                      (lis nil))
+                     ((or (null s) (graftp s) (eq s common-ancestor)) lis)
+                   (push s lis)))
+        (dispatch-event-copy s event 'pointer-enter-event)))))
+
+(defmethod distribute-event ((port basic-port) event)
+  (dispatch-event (event-sheet event) event))
+
+;;; In the most general case we can't tell whenever all sheets are
+;;; mirrored or not. So a default method for pointer-events operates
+;;; under the assumption that we must deliver events to sheets which
+;;; doesn't have a mirror and that the sheet grabbing and pressing is
```

Comment:
Maybe "that tracking of the grabbed sheet and pressed sheet is not ..."?

Reply:
I don't see how the full sentence would look like. "is not implemented locally"? or "is not implemented by port"? both statements are false: we do implement it locally and we overwrite whatever port does wrt grabbing/pressing etc.